### PR TITLE
Update Javadocs for repo source code

### LINKED ISSUE #8: Update Javadocs
With GH pages CI for Javadocs coming online, we should ensure that our Javadocs are defined appropriately for public end-user consumption.

### PR DESCRIPTION
**Pull request by CarlAmko**
Resolves #8

### CODE REVIEW

File: ngrok-java/src/main/java/com/ngrok/AgentTunnel.java [hunk 1]
```diff
@@ -3,70 +3,169 @@
 import java.util.ArrayList;
 import java.util.List;
 
+/**
+ * The `AgentTunnel` class is an abstract class that extends the {@link Tunnel}
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/AgentTunnel.java]: > The `AgentTunnel` is the base class for all agent initiated (e.g. configured by the agent) tunnels. Includes common attributes such as tunnel `protocol` and `url`.

🤔

File: ngrok-java/src/main/java/com/ngrok/AgentTunnel.java [hunk 2]
```diff
@@ -3,70 +3,169 @@
...
+ * class and represents a tunnel for an agent, with properties such as protocol,
+ * URL, and forwarding options.
+ * 
+ * @see {@link TcpTunnel}
+ * @see {@link TlsTunnel}
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/AgentTunnel.java]: `@see {@link HttpTunnel`}` probably in the first place 🤷

File: ngrok-java/src/main/java/com/ngrok/AgentTunnel.java [hunk 3]
```diff
@@ -3,70 +3,169 @@
...
+     * @param <T> the type of the builder
+     */
     public static abstract class Builder<T extends Builder> extends Tunnel.Builder<T> {
         private final List<String> allowCIDR = new ArrayList<>();
         private final List<String> denyCIDR = new ArrayList<>();
         private ProxyProto proxyProto;
         private String forwardsTo;
 
+        /**
+         * The function allows the addition of a CIDR (Classless Inter-Domain Routing)
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/AgentTunnel.java]: > Adds a CIDR to the list of allowed CIDRs

File: ngrok-java/src/main/java/com/ngrok/AgentTunnel.java [hunk 4]
```diff
@@ -3,70 +3,169 @@
...
+         * 
+         * @param allowCIDR The parameter "allowCIDR" is a string that represents a
+         *                  Classless
+         *                  Inter-Domain Routing (CIDR) notation. It is used to specify
+         *                  a range of IP addresses that are
+         *                  allowed.
+         * @return An instance of the class that the method is defined in, which is
+         *         represented by the generic type T.
+         */
         public T allowCIDR(String allowCIDR) {
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/AgentTunnel.java:71]: I wonder if we should add some abstraction/validation here 🤔

File: ngrok-java/src/main/java/com/ngrok/AgentTunnel.java [hunk 5]
```diff
@@ -3,70 +3,169 @@
...
+         * The function allows the addition of a CIDR (Classless Inter-Domain Routing)
+         * address to a
+         * list.
+         * 
+         * @param allowCIDR The parameter "allowCIDR" is a string that represents a
+         *                  Classless
+         *                  Inter-Domain Routing (CIDR) notation. It is used to specify
+         *                  a range of IP addresses that are
+         *                  allowed.
+         * @return An instance of the class that the method is defined in, which is
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/AgentTunnel.java]: ?

File: ngrok-java/src/main/java/com/ngrok/AgentTunnel.java [hunk 6]
```diff
@@ -3,70 +3,169 @@
...
+         *         represented by the generic type T.
+         */
         public T allowCIDR(String allowCIDR) {
             this.allowCIDR.add(allowCIDR);
             return (T) this;
         }
 
+        /**
+         * The function returns a list of strings representing allowed CIDR addresses.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/AgentTunnel.java]: > Returns the configured list of allowed CIDR addresses for this builder.

File: ngrok-java/src/main/java/com/ngrok/AgentTunnel.java [hunk 7]
```diff
@@ -3,70 +3,169 @@
...
+         * 
+         * @return A list of strings representing allowed CIDR addresses.
+         */
         public List<String> getAllowCIDR() {
             return allowCIDR;
         }
 
+        /**
+         * The function adds a denyCIDR to a list.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/AgentTunnel.java]: similar to the allow

File: ngrok-java/src/main/java/com/ngrok/AgentTunnel.java [hunk 8]
```diff
@@ -3,70 +3,169 @@
...
+         */
         public List<String> getDenyCIDR() {
             return denyCIDR;
         }
 
+        /**
+         * Sets the proxy protocol for the tunnel.
+         *
+         * @param proxyProto the proxy protocol for the tunnel
+         * @return the tunnel instance
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/AgentTunnel.java]: > returns the current builder instance

File: ngrok-java/src/main/java/com/ngrok/AgentTunnel.java [hunk 9]
```diff
@@ -3,70 +3,169 @@
...
+         * @return true if the tunnel has a proxy protocol, false otherwise
+         */
         public boolean hasProxyProto() {
             return proxyProto != null && proxyProto != ProxyProto.None;
         }
 
+        /**
+         * Returns the version of the proxy protocol for the tunnel.
+         *
+         * @return the version of the proxy protocol for the tunnel
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/AgentTunnel.java]: these are all about the builder, not the tunnel

File: ngrok-java/src/main/java/com/ngrok/Connection.java [hunk 10]
```diff
@@ -4,25 +4,62 @@
 import java.net.InetSocketAddress;
 import java.nio.ByteBuffer;
 
+/**
+ * Represents a generic socket connection to a remote target.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/Connection.java]: > Represents a connection established over this tunnel

File: ngrok-java/src/main/java/com/ngrok/Connection.java [hunk 11]
```diff
@@ -4,25 +4,62 @@
...
+     *
+     * @return {@link InetSocketAddress} representing the internet address
+     */
     public InetSocketAddress inetAddress() {
         var parts = remoteAddr.split(":");
         return new InetSocketAddress(parts[0], Integer.parseInt(parts[1]));
     }
 
+    /**
+     * Reads a sequence of bytes from this connection into the given buffer.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/Connection.java]: > Reads the next available bytes from this connection, up to the buffer capacity.

File: ngrok-java/src/main/java/com/ngrok/HttpTunnel.java [hunk 12]
```diff
@@ -5,378 +5,913 @@
 import java.util.Objects;
 import java.util.Optional;
 
+/**
+ * An abstract representation of a tunnel connection over HTTP.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/HttpTunnel.java]: 🤔 its not a tunnel connection over HTTP, but a tunnel that serves http traffic

File: ngrok-java/src/main/java/com/ngrok/HttpTunnel.java [hunk 13]
```diff
@@ -5,378 +5,913 @@
...
+     * @param metadata   the metadata of the tunnel
+     * @param proto      the protocol of the tunnel
+     * @param url        the URL of the tunnel
+     */
     public HttpTunnel(String id, String forwardsTo, String metadata, String proto, String url) {
         super(id, forwardsTo, metadata, proto, url);
     }
 
+    /**
+     * An enum representing the scheme of an HTTP tunnel.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/HttpTunnel.java]: > Represents the scheme for this tunnel

File: ngrok-java/src/main/java/com/ngrok/HttpTunnel.java [hunk 14]
```diff
@@ -5,378 +5,913 @@
...
         }
     }
 
+    /**
+     * A class representing OAuth options for an HTTP tunnel.
+     */
     public static class OAuthOptions {
+        /**
+         * The provider of the OAuth options.
+         */
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/HttpTunnel.java]: why are we documenting private fields?

File: ngrok-java/src/main/java/com/ngrok/HttpTunnel.java [hunk 15]
```diff
@@ -5,378 +5,913 @@
...
         private String allowDomain;
+
+        /**
+         * The scope of the OAuth options.
+         */
         private String scope;
 
+        /**
+         * Constructs a new set of OAuth options with the specified provider.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/HttpTunnel.java]: > Construct a new OAuth options with specific provider

File: ngrok-java/src/main/java/com/ngrok/HttpTunnel.java [hunk 16]
```diff
@@ -5,378 +5,913 @@
...
+         * @return this OAuthOptions object
+         */
         public OAuthOptions client(String id, String secret) {
             this.clientId = Objects.requireNonNull(id);
             this.clientSecret = Objects.requireNonNull(secret);
             return this;
         }
 
+        /**
+         * Sets the email address allowed by the OAuth options.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/HttpTunnel.java]: > Sets the email address allowed by OAuth

File: ngrok-java/src/main/java/com/ngrok/net/ConnectionOutputStream.java [hunk 17]
```diff
@@ -31,10 +57,18 @@ public void write(byte[] b, int off, int len) throws IOException {
...
+     * {@link #write(byte[], int, int)}.
+     *
+     * @throws IOException if an I/O error occurs
+     */
     @Override
     public void flush() throws IOException {
         buffer.flip();
         connection.write(buffer);
-        buffer.clear();
+        buffer.compact();
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/net/ConnectionOutputStream.java]: why this change?

File: ngrok-java/src/main/java/com/ngrok/net/ConnectionOutputStream.java [hunk 18]
```diff
@@ -6,22 +6,48 @@
...
+ * An output stream for writing data to an ngrok connection.
+ */
 public class ConnectionOutputStream extends OutputStream {
     private final Connection connection;
 
     private final ByteBuffer buffer;
 
+    /**
+     * Creates a new output stream for the given connection with the specified
+     * buffer size.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/net/ConnectionOutputStream.java]: expand on the buffer size, we are allocating a direct one?

File: ngrok-java/src/main/java/com/ngrok/AgentTunnel.java (hunk 4 above)
Comment (CarlAmko) [reply ngrok-java/src/main/java/com/ngrok/AgentTunnel.java:71]: Yeah that seems like a reasonable thing to do, though not as part of this PR.

File: ngrok-java/src/main/java/com/ngrok/HttpTunnel.java (hunk 14 above)
Comment (CarlAmko) [reply ngrok-java/src/main/java/com/ngrok/HttpTunnel.java]: It may not be necessary for Javadocs, but I don't see why it hurts?

File: ngrok-java-native/src/main/java/com/ngrok/NativeConnection.java [hunk 19]
```diff
@@ -3,13 +3,28 @@
 import java.io.IOException;
 import java.nio.ByteBuffer;
 
+/**
+ * A class representing a connection for an HTTP tunnel with native implementation.
```
Comment (nikolay-ngrok) [inline ngrok-java-native/src/main/java/com/ngrok/NativeConnection.java]: > An implementation of {link Connection} that delegates calls to a native library.

or something along these lines, probably for most of the `Native*` classes.

File: ngrok-java-native/src/main/java/com/ngrok/NativeHttpTunnel.java [hunk 20]
```diff
@@ -1,19 +1,50 @@
...
+     * @return a new NativeConnection object representing the accepted connection
+     * @throws IOException if an I/O error occurs
+     */
     @Override
     public native NativeConnection accept() throws IOException;
 
+    /**
+     * Forwards TCP traffic to the specified address.
+     *
+     * @param addr the address to forward TCP traffic to
```
Comment (nikolay-ngrok) [inline ngrok-java-native/src/main/java/com/ngrok/NativeHttpTunnel.java]: add what form is the address expected to be

File: ngrok-java-native/src/main/java/com/ngrok/NativeHttpTunnel.java [hunk 21]
```diff
@@ -1,19 +1,50 @@
...
+     * @param metadata   the metadata of the tunnel
+     * @param proto      the protocol of the tunnel
+     * @param url        the URL of the tunnel
+     */
     public NativeHttpTunnel(String id, String forwardsTo, String metadata, String proto, String url) {
         super(id, forwardsTo, metadata, proto, url);
     }
 
+    /**
+     * Accepts a new connection on the HTTP tunnel.
```
Comment (nikolay-ngrok) [inline ngrok-java-native/src/main/java/com/ngrok/NativeHttpTunnel.java]: we should add its blocking until one is available.

File: ngrok-java-native/src/main/java/com/ngrok/NativeHttpTunnel.java [hunk 22]
```diff
@@ -1,19 +1,50 @@
...
+     *
+     * @return a new NativeConnection object representing the accepted connection
+     * @throws IOException if an I/O error occurs
+     */
     @Override
     public native NativeConnection accept() throws IOException;
 
+    /**
+     * Forwards TCP traffic to the specified address.
```
Comment (nikolay-ngrok) [inline ngrok-java-native/src/main/java/com/ngrok/NativeHttpTunnel.java]: should add its blocked until closed or something 🤷

File: ngrok-java-native/src/main/java/com/ngrok/NativeSession.java [hunk 23]
```diff
@@ -26,32 +29,90 @@ public class NativeSession implements Session {
...
+     * Constructs a new native session with the specified metadata.
+     *
+     * @param metadata the metadata of the session
+     */
     public NativeSession(String metadata) {
         this.metadata = metadata;
     }
 
+    /**
+     * Connects to an ngrok tunnel using the specified session builder.
```
Comment (nikolay-ngrok) [inline ngrok-java-native/src/main/java/com/ngrok/NativeSession.java]: > Establishes an ngrok session specified by the provided builder.

File: ngrok-java-native/src/main/java/com/ngrok/NativeSession.java [hunk 24]
```diff
@@ -26,32 +29,90 @@ public class NativeSession implements Session {
...
+     * @return a new NativeSession object representing the connected session
+     * @throws IOException if an I/O error occurs
+     */
     public static NativeSession connect(Session.Builder builder) throws IOException {
         builder.getUserAgents().add(0, new UserAgent("ngrok-java", version));
         return connectNative(builder);
     }
 
+    /**
+     * Connects to an ngrok tunnel using the specified session builder using native
```
Comment (nikolay-ngrok) [inline ngrok-java-native/src/main/java/com/ngrok/NativeSession.java]: similar

File: ngrok-java/src/main/java/com/ngrok/AgentTunnel.java [hunk 25]
```diff
@@ -3,70 +3,160 @@
 import java.util.ArrayList;
 import java.util.List;
 
+/**
+ * The AgentTunnel is the base class for all agent initiated (e.g. configured by the agent) tunnels.
+ * Includes common attributes such as tunnel protocol and url.
+ * 
+ * @see {@link HttpTunnel}
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/AgentTunnel.java]:
```
 * @see {@link TcpTunnel}
 * @see {@link TlsTunnel}
```

File: ngrok-java/src/main/java/com/ngrok/net/AbstractSocketImpl.java [hunk 26]
```diff
@@ -8,72 +8,117 @@
...
+    /**
+     * See {@link SocketImpl#setOption(int, Object)}
+     */
     @Override
     public void setOption(int optID, Object value) throws SocketException {
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * See {@link SocketImpl#getOption(int)}
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/net/AbstractSocketImpl.java:120]: that's what I would have expected by most `Native*` classes 🤷

File: ngrok-java/src/main/java/com/ngrok/Session.java [hunk 27]
```diff
@@ -7,16 +7,40 @@
 import java.util.List;
 import java.util.Objects;
 
+/**
+ * A session with the ngrok service.
+ */
 public interface Session extends AutoCloseable {
 
+    /**
+     * Returns a new {@link Builder} instance with your ngrok authentication token
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/Session.java]: > Creates a new {@link Builder} instance. The ngrok authtoken will be set to the value of the `NGROK_AUTHOKEN` environment variable.

File: ngrok-java/src/main/java/com/ngrok/Session.java [hunk 28]
```diff
@@ -7,16 +7,40 @@
...
+     * from the environment.
+     *
+     * @return a new {@link Builder} instance with the default authentication token
+     */
     public static Builder newBuilder() {
         return newBuilder(System.getenv("NGROK_AUTHTOKEN"));
     }
 
+    /**
+     * Returns a new {@link Builder} instance with the specified ngrok
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/Session.java]: > Creates a new {@link Builder} instance with specified authtoken.

File: ngrok-java/src/main/java/com/ngrok/Session.java [hunk 29]
```diff
@@ -33,47 +57,131 @@ public static Session connect(Builder builder) throws IOException {
...
+     * Returns a new {@link TcpTunnel} instance with the default settings
+     * 
+     * @throws IOException if an I/O error occurs during the tunnel creation
+     */
     public default TcpTunnel tcpTunnel() throws IOException {
         return tcpTunnel(new TcpTunnel.Builder());
     }
 
+    /**
+     * Returns a new {@link TcpTunnel} instance with the specified settings.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/Session.java]: We should expand on these - it doesn't just create a new instance, it actually creates a new tunnel and returns it.

File: ngrok-java/src/main/java/com/ngrok/Session.java [hunk 30]
```diff
@@ -33,47 +57,131 @@ public static Session connect(Builder builder) throws IOException {
...
+     *
+     * @param builder the {@link LabeledTunnel.Builder} instance to use for the tunnel
+     *                creation
+     * @return a new {@link LabeledTunnel} instance with the specified settings
+     * @throws IOException if an I/O error occurs during the tunnel creation
+     */
     public LabeledTunnel labeledTunnel(LabeledTunnel.Builder builder) throws IOException;
 
+    /**
+     * Represents a callback for stopping a session. Implements
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/Session.java]: > A callback used when the agent receives stop session from the ngrok server.   

or maybe just copy the first paragraph from: https://docs.rs/ngrok/latest/ngrok/session/struct.SessionBuilder.html#method.handle_stop_command ? (also for the rest callbacks)

File: ngrok-java/src/main/java/com/ngrok/TcpTunnel.java [hunk 31]
```diff
@@ -2,25 +2,59 @@
 
 import java.util.Objects;
 
+/**
+ * Uses an agent-aware tunnel to create a TCP connection to ngrok.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/TcpTunnel.java]: > Represents a tcp tunnel with the ngrok service.

File: ngrok-java/src/main/java/com/ngrok/TlsTunnel.java [hunk 32]
```diff
@@ -2,11 +2,29 @@
 
 import java.util.Objects;
 
+/**
+ * Uses an agent-aware tunnel to create a TLS connection to ngrok.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/TlsTunnel.java]: change wording

File: ngrok-java/src/main/java/com/ngrok/TlsTunnel.java [hunk 33]
```diff
@@ -15,38 +33,82 @@ public static class Builder extends AgentTunnel.Builder<Builder> {
...
+         * Returns the termination key in PEM format.
+         *
+         * @return the termination key in PEM format
+         */
         public byte[] getTerminationKeyPEM() {
             return terminationKeyPEM;
         }
 
+        /**
+         * Checks if the domain is not null.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/TlsTunnel.java]: > Checks if the domain is set

File: ngrok-java/src/main/java/com/ngrok/Tunnel.java [hunk 34]
```diff
@@ -2,49 +2,114 @@
 
 import java.io.IOException;
 
+/**
+ * Represents an abstract ngrok tunnel. Can be expanded upon to define new ways to create tunnel connections.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/Tunnel.java]: just the first sentence?

> Base class for all types of ngrok tunnels. Includes common attributes and functionality shared by all types of tunnels
> See..

**Review by nikolay-ngrok — COMMENTED** [review]
some additional changes 👍

File: ngrok-java/src/main/java/com/ngrok/LabeledTunnel.java [hunk 35]
```diff
@@ -6,41 +6,87 @@
 import java.util.Objects;
 import java.util.stream.Collectors;
 
+/**
+ * A subclass of {@link Tunnel} that additionally tracks labels as key/value pairs.
```
Comment (nikolay-ngrok) [inline ngrok-java/src/main/java/com/ngrok/LabeledTunnel.java]: > Represents a labeled tunnel with the ngrok service. Labeled tunnels are configured by using the edges functionality [link]
?

**Review by nikolay-ngrok — APPROVED** [review]
👍 had a comment pending, hopefully you can address it
